packages/fuels-core/src/types/param_types.rs

```diff
@@ -84,57 +84,85 @@ impl ParamType {
-            ParamType::Vector(param_type) => param_type.uses_heap_types(),
-            ParamType::Bytes => false,
-            // Here, we return false because even though the `Token::String` type has an underlying
-            // `Bytes` type nested, it is an exception that will be generalized as part of
-            // https://github.com/FuelLabs/fuels-rs/discussions/944
-            ParamType::String => false,
-            _ => self.uses_heap_types(),
-        }
-    }
-
-    fn uses_heap_types(&self) -> bool {
-        match &self {
-            ParamType::Vector(..) | ParamType::Bytes | ParamType::String => true,
-            ParamType::Array(param_type, ..) => param_type.uses_heap_types(),
-            ParamType::Tuple(param_types, ..) => Self::any_nested_heap_types(param_types),
-            ParamType::Enum {
-                generics, variants, ..
-            } => {
-                let variants_types = variants.param_types();
-                Self::any_nested_heap_types(chain!(generics, variants_types))
+    pub fn children_need_extra_receipts(&self) -> bool {
+        match self {
+            ParamType::Array(inner, _) | ParamType::Vector(inner) => {
+                inner.needs_extra_data_receipt(false)
             }
-            ParamType::Struct {
-                fields, generics, ..
-            } => Self::any_nested_heap_types(chain!(fields, generics)),
+            ParamType::Struct { fields, .. } => fields
+                .iter()
+                .any(|param_type| param_type.needs_extra_data_receipt(false)),
+            ParamType::Enum { variants, .. } => variants
+                .param_types()
+                .iter()
+                .any(|param_type| param_type.needs_extra_data_receipt(false)),
+            ParamType::Tuple(inner_types) => inner_types
+                .iter()
+                .any(|param_type| param_type.needs_extra_data_receipt(false)),
             _ => false,
         }
     }
 
-    fn any_nested_heap_types<'a>(param_types: impl IntoIterator<Item = &'a ParamType>) -> bool {
-        param_types
-            .into_iter()
-            .any(|param_type| param_type.uses_heap_types())
+    pub fn validate_is_decodable(&self) -> Result<()> {
+        match self {
+            ParamType::Enum { variants, .. } => {
+                let all_param_types = chain!(variants.param_types());
+                let grandchildren_need_receipts = all_param_types
+                    .clone()
+                    .any(|child| child.children_need_extra_receipts());
+                if grandchildren_need_receipts {
+                    return Err(error!(
+                        InvalidType,
+                        "Enums currently support only one level deep heap types."
+                    ));
+                }
+
+                let num_of_children_needing_receipts = all_param_types
+                    .filter(|param_type| param_type.needs_extra_data_receipt(false))
+                    .count();
```

Comment:
No need for `chain!` and cloning of the param types:
```suggestion
                let all_param_types = variants.param_types();
                let grandchildren_need_receipts = all_param_types
                    .iter()
                    .any(|child| child.children_need_extra_receipts());
                if grandchildren_need_receipts {
                    return Err(error!(
                        InvalidType,
                        "Enums currently support only one level deep heap types."
                    ));
                }

                let num_of_children_needing_receipts = all_param_types
                    .iter()
                    .filter(|param_type| param_type.needs_extra_data_receipt(false))
                    .count();
```